File: arrays/python/count-inversions.py
"""
Algorithm Type: Divide & Conquer
Time Complexity: O(n*log(n))
"""
numbers = [8, 2, 1, 5, 7, 3, 9, 2, 0, 1]


def count_split_inv(arr, left, right):
  split_inv = ridx = lidx = 0
  size = len(arr)
  lsize = len(left)
  rsize = len(right)
  for i in range(size):
    if lidx != lsize and ridx != rsize:
      if right[ridx] < left[lidx]:
        arr[i] = right[ridx]
        ridx += 1
        split_inv += lsize - lidx
      else:
        arr[i] = left[lidx]
        lidx += 1
    elif lidx == lsize:
      arr[i] = right[ridx]
      ridx += 1
    elif ridx == rsize:
      arr[i] = left[lidx]
      lidx += 1
  return split_inv


def count_inversions(arr) -> int:
  """
  Sort arr and return the number of inversions required.

  >>> numbers
  [8, 2, 1, 5, 7, 3, 9, 2, 0, 1]
  >>> count_inversions(numbers)
  28
  >>> numbers  # numbers has been sorted!!
  [0, 1, 1, 2, 2, 3, 5, 7, 8, 9]
  >>> count_inversions(numbers)
  0
  """
  size = len(arr)
  if size == 1:
    return 0
  mid = int(size/2)
  left = arr[:mid]
  right = arr[mid:]
  l_inv = count_inversions(left)
  r_inv = count_inversions(right)
  split_inv = count_split_inv(arr, left, right)

  return l_inv + r_inv + split_inv

if __name__ == "__main__":
  print(count_inversions(numbers))

File: arrays/java/count-inversions.java
// Algorithm Type: Divide & Conquer
// Time Complexity: O(n*log(n))

import java.io.File;
import java.lang.reflect.Array;
import java.util.*;

public class inversions {
  private static long count_split_inv(int[] arr, int[] left, int[] right) {
    long split_inv = 0;
    int ridx = 0, lidx = 0;
    int size = arr.length;
    int rsize = right.length;
    int lsize = left.length;
    for (int i = 0; i < size; i++) {
      if (lidx != lsize && ridx != rsize) {
        if (right[ridx] <= left[lidx]) {
          arr[i] = right[ridx];
          ridx++;
          split_inv += lsize - lidx;
        } else {
          arr[i] = left[lidx];
          lidx++;
        }
      } else if (lidx == lsize) {
        arr[i] = right[ridx];
        ridx++;
      } else if (ridx == rsize) {
        arr[i] = left[lidx];
        lidx++;
      }
    }
    return split_inv;
  }

  private static long count_inversions(int[] arr) {
    int size = arr.length;
    if (size == 1) {
      return 0;
    }
    int[] left = Arrays.copyOfRange(arr, 0, size / 2);
    int[] right = Arrays.copyOfRange(arr, size / 2, size);
    long left_inv = count_inversions(left);
    long right_inv = count_inversions(right);
    long split_inv = count_split_inv(arr, left, right);

    return left_inv + right_inv + split_inv;
  }

  public static void main(String[] args) {
    try {
      int[] arr = {8, 2, 1, 5, 7, 3, 9, 2, 0, 1};
      System.out.println(count_inversions(arr));
    } catch (Exception e) {
      System.out.println("Err... ");
    }

  }
}
